File: iOS/Pulsar/Sources/Pulsar/Presets/PresetsWrapper.swift
import AVFoundation
import UIKit
import Foundation
import AVFAudio

@objc public class PresetsWrapper : NSObject {
  private var playSound: Bool = true
  private var useCache: Bool = true
  private var cache: [String: Preset] = [:]
  private var haptics: Pulsar!
  
  private var mapper: [String: Preset.Type] = [
    "Earthquake": EarthquakePreset.self,
    "Success": SuccessPreset.self,
    "Fail": FailPreset.self,
    "Tap": TapPreset.self,
    "SystemImpactLight": SystemImpactLightPreset.self,
    "SystemImpactMedium": SystemImpactMediumPreset.self,
    "SystemImpactHeavy": SystemImpactHeavyPreset.self,
    "SystemImpactSoft": SystemImpactSoftPreset.self,
    "SystemImpactRigid": SystemImpactRigidPreset.self,
    "SystemImpactSuccess": SystemNotificationSuccessPreset.self,
    "SystemNotificationWarning": SystemNotificationWarningPreset.self,
    "SystemNotificationError": SystemNotificationErrorPreset.self,
    "SystemSelection": SystemSelectionPreset.self,
  ]
  
  public init(haptics: Pulsar) {
    super.init()
    self.haptics = haptics
  }
  
  public func enableCache(state: Bool) {
    self.useCache = state
    if (!state) {
      resetCache()
    }
  }
  
  public func isCacheEnabled() -> Bool {
    return self.useCache
  }
  
  public func resetCache() {
    cache.removeAll()
  }
  
  public func preloadPresetByName(_ name: String) {
    self.useCache = true
    _ = getCacheablePreset(mapper[name]!)
  }
  
  @objc public func getByName(_ name: String) -> Preset? {
    guard mapper.keys.contains(name) else {
      return nil
    }
    return getCacheablePreset(mapper[name]!)
  }
  
  private func getCacheablePreset(_ type: Preset.Type) -> Preset {
    if (useCache) {
      if let cachedPreset = cache[type.name] {
        return cachedPreset
      } else {
        let preset = type.getInstance(haptics: haptics!)
        cache[type.name] = preset
        return preset
      }
    }
    return type.getInstance(haptics: haptics!)
  }
  
  @objc public func Earthquake() {
    getCacheablePreset(EarthquakePreset.self).play()
  }

  @objc public func Success() {
    getCacheablePreset(SuccessPreset.self).play()
  }

  @objc public func Fail() {
    getCacheablePreset(FailPreset.self).play()
  }
  
  @objc public func Tap() {
    getCacheablePreset(TapPreset.self).play()
  }
  
  @objc public func SystemImpactLight() {
    getCacheablePreset(SystemImpactLightPreset.self).play()
  }
  
  @objc public func SystemImpactMedium() {
    getCacheablePreset(SystemImpactMediumPreset.self).play()
  }
  
  @objc public func SystemImpactHeavy() {
    getCacheablePreset(SystemImpactHeavyPreset.self).play()
  }
  
  @objc public func SystemImpactSoft() {
    getCacheablePreset(SystemImpactSoftPreset.self).play()
  }
  
  @objc public func SystemImpactRigid() {
    getCacheablePreset(SystemImpactRigidPreset.self).play()
  }
  
  @objc public func SystemNotificationSuccess() {
    getCacheablePreset(SystemNotificationSuccessPreset.self).play()
  }
  
  @objc public func SystemNotificationWarning() {
    getCacheablePreset(SystemNotificationWarningPreset.self).play()
  }
  
  @objc public func SystemNotificationError() {
    getCacheablePreset(SystemNotificationErrorPreset.self).play()
  }
  
  @objc public func SystemSelection() {
    getCacheablePreset(SystemSelectionPreset.self).play()
  }
}

File: react-native/react-native-pulsar/android/src/main/java/com/swmansion/pulsar/presets/PresetsWrapper.kt
package com.swmansion.pulsar.presets

import android.content.Context
import com.swmansion.pulsar.Pulsar
import com.swmansion.pulsar.haptics.HapticEngineWrapper
import com.swmansion.pulsar.types.Preset

class PresetsWrapper(
    private val haptics: Pulsar,
    context: Context,
    engine: HapticEngineWrapper,
) {
    private var useCache: Boolean = true
    private val cache = mutableMapOf<String, Preset>()
    private val systemPrimitivePresets = SystemPrimitivePresets(engine)
    private val systemViewBasedPresets = SystemViewBasedPresets(context)

    private val mapper: Map<String, (Pulsar) -> Preset> = mapOf(
        EarthquakePreset.name to { haptics -> EarthquakePreset(haptics) },
        SuccessPreset.name to { haptics -> SuccessPreset(haptics) },
        FailPreset.name to { haptics -> FailPreset(haptics) },
        TapPreset.name to { haptics -> TapPreset(haptics) },

        SystemImpactLightPreset.name to { haptics -> SystemImpactLightPreset(haptics) },
        SystemImpactMediumPreset.name to { haptics -> SystemImpactMediumPreset(haptics) },
        SystemImpactHeavyPreset.name to { haptics -> SystemImpactHeavyPreset(haptics) },
        SystemImpactSoftPreset.name to { haptics -> SystemImpactSoftPreset(haptics) },
        SystemImpactRigidPreset.name to { haptics -> SystemImpactRigidPreset(haptics) },
        SystemNotificationSuccessPreset.name to { haptics -> SystemNotificationSuccessPreset(haptics) },
        SystemNotificationWarningPreset.name to { haptics -> SystemNotificationWarningPreset(haptics) },
        SystemNotificationErrorPreset.name to { haptics -> SystemNotificationErrorPreset(haptics) },

        SystemEffectClickPreset.name to { haptics -> SystemEffectClickPreset(haptics, systemPrimitivePresets) },
        SystemDoubleClickPreset.name to { haptics -> SystemDoubleClickPreset(haptics, systemPrimitivePresets) },
        SystemTickPreset.name to { haptics -> SystemTickPreset(haptics, systemPrimitivePresets) },
        SystemHeavyClickPreset.name to { haptics -> SystemHeavyClickPreset(haptics, systemPrimitivePresets) },

        SystemLongPressPreset.name to { haptics -> SystemLongPressPreset(haptics, systemViewBasedPresets) },
        SystemVirtualKeyPreset.name to { haptics -> SystemVirtualKeyPreset(haptics, systemViewBasedPresets) },
        SystemKeyboardTapPreset.name to { haptics -> SystemKeyboardTapPreset(haptics, systemViewBasedPresets) },
        SystemClockTickPreset.name to { haptics -> SystemClockTickPreset(haptics, systemViewBasedPresets) },
        SystemCalendarDatePreset.name to { haptics -> SystemCalendarDatePreset(haptics, systemViewBasedPresets) },
        SystemContextClickPreset.name to { haptics -> SystemContextClickPreset(haptics, systemViewBasedPresets) },
        SystemKeyboardPressPreset.name to { haptics -> SystemKeyboardPressPreset(haptics, systemViewBasedPresets) },
        SystemKeyboardReleasePreset.name to { haptics -> SystemKeyboardReleasePreset(haptics, systemViewBasedPresets) },
        SystemVirtualKeyReleasePreset.name to { haptics -> SystemVirtualKeyReleasePreset(haptics, systemViewBasedPresets) },
        SystemTextHandleMovePreset.name to { haptics -> SystemTextHandleMovePreset(haptics, systemViewBasedPresets) },
        SystemDragCrossingPreset.name to { haptics -> SystemDragCrossingPreset(haptics, systemViewBasedPresets) },
        SystemGestureStartPreset.name to { haptics -> SystemGestureStartPreset(haptics, systemViewBasedPresets) },
        SystemGestureEndPreset.name to { haptics -> SystemGestureEndPreset(haptics, systemViewBasedPresets) },
        SystemEdgeSqueezePreset.name to { haptics -> SystemEdgeSqueezePreset(haptics, systemViewBasedPresets) },
        SystemEdgeReleasePreset.name to { haptics -> SystemEdgeReleasePreset(haptics, systemViewBasedPresets) },
        SystemConfirmPreset.name to { haptics -> SystemConfirmPreset(haptics, systemViewBasedPresets) },
        SystemReleasePreset.name to { haptics -> SystemReleasePreset(haptics, systemViewBasedPresets) },
        SystemScrollTickPreset.name to { haptics -> SystemScrollTickPreset(haptics, systemViewBasedPresets) },
        SystemScrollItemFocusPreset.name to { haptics -> SystemScrollItemFocusPreset(haptics, systemViewBasedPresets) },
        SystemScrollLimitPreset.name to { haptics -> SystemScrollLimitPreset(haptics, systemViewBasedPresets) },
        SystemToggleOnPreset.name to { haptics -> SystemToggleOnPreset(haptics, systemViewBasedPresets) },
        SystemToggleOffPreset.name to { haptics -> SystemToggleOffPreset(haptics, systemViewBasedPresets) },
        SystemDragStartPreset.name to { haptics -> SystemDragStartPreset(haptics, systemViewBasedPresets) },
        SystemSegmentTickPreset.name to { haptics -> SystemSegmentTickPreset(haptics, systemViewBasedPresets) },
        SystemSegmentFrequentTickPreset.name to { haptics -> SystemSegmentFrequentTickPreset(haptics, systemViewBasedPresets) },
    )

    fun enableCache(state: Boolean) {
        this.useCache = state
        if (!state) {
            resetCache()
        }
    }

    fun isCacheEnabled(): Boolean = this.useCache

    fun resetCache() {
        cache.clear()
    }

    fun preloadPresetByName(name: String) {
        this.useCache = true
        getCacheablePreset(name)
    }

    fun getByName(name: String): Preset? {
        return getCacheablePreset(name)
    }

    private fun getCacheablePreset(name: String): Preset? {
        return if (useCache) {
            cache.getOrPut(name) {
                mapper[name]?.invoke(haptics) ?: return null
            }
        } else {
            mapper[name]?.invoke(haptics) ?: return null
        }
    }

    fun Earthquake() {
        getCacheablePreset(EarthquakePreset.name)!!.play()
    }

    fun Success() {
        getCacheablePreset(SuccessPreset.name)!!.play()
    }

    fun Fail() {
        getCacheablePreset(FailPreset.name)!!.play()
    }

    fun Tap() {
        getCacheablePreset(TapPreset.name)!!.play()
    }

    
}
